File: samlang-checker/src/test/kotlin/samlang/checker/DependencyTrackerTest.kt
package samlang.checker

import io.kotlintest.shouldBe
import io.kotlintest.specs.StringSpec
import samlang.ast.common.ModuleReference

class DependencyTrackerTest : StringSpec() {
    init {
        "can track and update dependencies" {
            val tracker = DependencyTracker()
            val moduleA = ModuleReference(moduleName = "A")
            val moduleB = ModuleReference(moduleName = "B")
            val moduleC = ModuleReference(moduleName = "C")
            val moduleD = ModuleReference(moduleName = "D")
            val moduleE = ModuleReference(moduleName = "E")
            // Wave 1
            tracker.update(moduleReference = moduleA, importedModules = listOf(moduleB, moduleC))
            tracker.update(moduleReference = moduleD, importedModules = listOf(moduleB, moduleC))
            tracker.update(moduleReference = moduleE, importedModules = listOf(moduleB, moduleC))
            tracker.getForwardDependencies(moduleReference = moduleA) shouldBe setOf(moduleB, moduleC)
            tracker.getForwardDependencies(moduleReference = moduleB) shouldBe emptySet()
            tracker.getForwardDependencies(moduleReference = moduleC) shouldBe emptySet()
            tracker.getForwardDependencies(moduleReference = moduleD) shouldBe setOf(moduleB, moduleC)
            tracker.getForwardDependencies(moduleReference = moduleE) shouldBe setOf(moduleB, moduleC)
            tracker.getReverseDependencies(moduleReference = moduleA) shouldBe emptySet()
            tracker.getReverseDependencies(moduleReference = moduleB) shouldBe setOf(moduleA, moduleD, moduleE)
            tracker.getReverseDependencies(moduleReference = moduleC) shouldBe setOf(moduleA, moduleD, moduleE)
            tracker.getReverseDependencies(moduleReference = moduleD) shouldBe emptySet()
            tracker.getReverseDependencies(moduleReference = moduleE) shouldBe emptySet()
            // Wave 2
            tracker.update(moduleReference = moduleA, importedModules = listOf(moduleD, moduleE))
            tracker.getForwardDependencies(moduleReference = moduleA) shouldBe setOf(moduleD, moduleE)
            tracker.getForwardDependencies(moduleReference = moduleB) shouldBe emptySet()
            tracker.getForwardDependencies(moduleReference = moduleC) shouldBe emptySet()
            tracker.getForwardDependencies(moduleReference = moduleD) shouldBe setOf(moduleB, moduleC)
            tracker.getForwardDependencies(moduleReference = moduleE) shouldBe setOf(moduleB, moduleC)
            tracker.getReverseDependencies(moduleReference = moduleA) shouldBe emptySet()
            tracker.getReverseDependencies(moduleReference = moduleB) shouldBe setOf(moduleD, moduleE)
            tracker.getReverseDependencies(moduleReference = moduleC) shouldBe setOf(moduleD, moduleE)
            tracker.getReverseDependencies(moduleReference = moduleD) shouldBe setOf(moduleA)
            tracker.getReverseDependencies(moduleReference = moduleE) shouldBe setOf(moduleA)
            // Wave 3
            tracker.update(moduleReference = moduleA, importedModules = null)
            tracker.getForwardDependencies(moduleReference = moduleA) shouldBe emptySet()
            tracker.getForwardDependencies(moduleReference = moduleB) shouldBe emptySet()
            tracker.getForwardDependencies(moduleReference = moduleC) shouldBe emptySet()
            tracker.getForwardDependencies(moduleReference = moduleD) shouldBe setOf(moduleB, moduleC)
            tracker.getForwardDependencies(moduleReference = moduleE) shouldBe setOf(moduleB, moduleC)
            tracker.getReverseDependencies(moduleReference = moduleA) shouldBe emptySet()
            tracker.getReverseDependencies(moduleReference = moduleB) shouldBe setOf(moduleD, moduleE)
            tracker.getReverseDependencies(moduleReference = moduleC) shouldBe setOf(moduleD, moduleE)
            tracker.getReverseDependencies(moduleReference = moduleD) shouldBe setOf()
            tracker.getReverseDependencies(moduleReference = moduleE) shouldBe setOf()
        }
    }
}

File: samlang-core/checker/__tests/dependency-tracker.test.ts
import DependencyTracker from '../dependency-tracker';

import { ModuleReference } from 'samlang-core-ast/common-nodes';

it('can track and update dependencies', () => {
  const tracker = new DependencyTracker();
  const moduleA = new ModuleReference(['A']);
  const moduleB = new ModuleReference(['B']);
  const moduleC = new ModuleReference(['C']);
  const moduleD = new ModuleReference(['D']);
  const moduleE = new ModuleReference(['E']);

  // Wave 1
  tracker.update(moduleA, [moduleB, moduleC]);
  tracker.update(moduleD, [moduleB, moduleC]);
  tracker.update(moduleE, [moduleB, moduleC]);
  tracker.update(moduleA, [moduleB, moduleC]);
  tracker.update(moduleD, [moduleB, moduleC]);
  tracker.update(moduleE, [moduleB, moduleC]);
  expect(tracker.getForwardDependencies(moduleA).toArray()).toEqual([moduleB, moduleC]);
  expect(tracker.getForwardDependencies(moduleB).toArray()).toEqual([]);
  expect(tracker.getForwardDependencies(moduleC).toArray()).toEqual([]);
  expect(tracker.getForwardDependencies(moduleD).toArray()).toEqual([moduleB, moduleC]);
  expect(tracker.getForwardDependencies(moduleE).toArray()).toEqual([moduleB, moduleC]);
  expect(tracker.getReverseDependencies(moduleA).toArray()).toEqual([]);
  expect(tracker.getReverseDependencies(moduleB).toArray()).toEqual([moduleA, moduleD, moduleE]);
  expect(tracker.getReverseDependencies(moduleC).toArray()).toEqual([moduleA, moduleD, moduleE]);
  expect(tracker.getReverseDependencies(moduleD).toArray()).toEqual([]);
  expect(tracker.getReverseDependencies(moduleE).toArray()).toEqual([]);

  // Wave 2
  tracker.update(moduleA, [moduleD, moduleE]);
  expect(tracker.getForwardDependencies(moduleA).toArray()).toEqual([moduleD, moduleE]);
  expect(tracker.getForwardDependencies(moduleB).toArray()).toEqual([]);
  expect(tracker.getForwardDependencies(moduleC).toArray()).toEqual([]);
  expect(tracker.getForwardDependencies(moduleD).toArray()).toEqual([moduleB, moduleC]);
  expect(tracker.getForwardDependencies(moduleE).toArray()).toEqual([moduleB, moduleC]);
  expect(tracker.getReverseDependencies(moduleA).toArray()).toEqual([]);
  expect(tracker.getReverseDependencies(moduleB).toArray()).toEqual([moduleD, moduleE]);
  expect(tracker.getReverseDependencies(moduleC).toArray()).toEqual([moduleD, moduleE]);
  expect(tracker.getReverseDependencies(moduleD).toArray()).toEqual([moduleA]);
  expect(tracker.getReverseDependencies(moduleE).toArray()).toEqual([moduleA]);

  // Wave 3
  tracker.update(moduleA);
  expect(tracker.getForwardDependencies(moduleA).toArray()).toEqual([]);
  expect(tracker.getForwardDependencies(moduleB).toArray()).toEqual([]);
  expect(tracker.getForwardDependencies(moduleC).toArray()).toEqual([]);
  expect(tracker.getForwardDependencies(moduleD).toArray()).toEqual([moduleB, moduleC]);
  expect(tracker.getForwardDependencies(moduleE).toArray()).toEqual([moduleB, moduleC]);
  expect(tracker.getReverseDependencies(moduleA).toArray()).toEqual([]);
  expect(tracker.getReverseDependencies(moduleB).toArray()).toEqual([moduleD, moduleE]);
  expect(tracker.getReverseDependencies(moduleC).toArray()).toEqual([moduleD, moduleE]);
  expect(tracker.getReverseDependencies(moduleD).toArray()).toEqual([]);
  expect(tracker.getReverseDependencies(moduleE).toArray()).toEqual([]);
});
